flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java

```diff
@@ -51,6 +53,9 @@
     /** Executor for async state deletion */
     private final Executor asyncDisposalExecutor;
 
+    /** Checkpoint ID below which no state is discarded, inclusive. */
+    private long highestRetainCheckpointID = -1L;
```

Comment:
I think `highestRetainCheckpointID` might not be a good choice as we could still retain multi checkpoints even in `CLAIM` restore mode. However, current implementation would still keep the `highestRetainCheckpointID` as `-1` in `CLAIM` restore mode.

Reply:
Could you explain which case do you mean and what would be the problem?

With `CLAIM` mode, the goal is to discard the state, including retained checkpoints. So this field `highestRetainCheckpointID` must be `-1`, which will allow to do this (condition in state entry will be met).

Reply:
I feel that the naming of `retain` could make user mixed with the configuration `state.checkpoints.num-retained`. How about `reservedCheckpointIdOnRestore`, and we give clear description that `-1` means no checkpoint needs to be reserved on restore (actualy CLAIM mode).